### PR TITLE
Crop request

### LINKED ISSUE #602: Improve "request new crops" process
As an ordinary Growstuff member, I want to be able to request crops by some mechanism other than posting in a forum.

Specifically, I should be able to use a variant of the "new crop" form, which asks me for as much detail as I have, then records that crop as "requested" or something but doesn't make it live until reviewed by a crop wrangler.

When a crop is requested, email should be sent to all crop wranglers letting them know about it.

As a crop wrangler, I should be able to see these pending crops on the crop wrangler page, and approve them (making edits if necessary).  Alternately, I should be able to reject a crop, providing a reason (eg. "not edible").  It would be good to have a list of standard rejection reasons to select from plus "other".

When the crop is approved, an email should be sent to the requester letting them know it's been added and providing links to plant, harvest, or stash seeds.  If rejected, an email should be sent with the reason.

Some notes:

1) We might want to ease up on validation a bit, eg. to allow them to request it even if they don't know the Wikipedia link or whatever

2) Be careful of race conditions if multiple crop wranglers both come to deal with a new crop simultaneously.

3) We'll need some discussion about how to actually track the state of the crop. I'd suggest that we simply have a "status" field which can be "requested", "approved", and "rejected". All current crops should be set to "approved".

4) We'll also need a requested_by field to track who asked for a crop so we can email them.

5) Keep in mind that scientific names will need to be managed in similar ways.

6) We'll want to modify the help text on crops/new to be targetted at the general public vs wranglers. OTOH, crops/edit should be wrangler-specific.

### PR DESCRIPTION
**Pull request by tygriffin**
Fixes #602

**WIP: Needs feedback**

1. Is the language natural enough for regular members on the new crops form page?

2. Should we use Javascript to pop up the reason for rejection only when `approval_status` is "rejected" on the edit crops form?

3. Is the `reason_for_rejection` required if rejected?

4. Should the reasons for rejection be hardcoded in the model or should they be in a database table? (Database seems to make more sense. That way, if a wrangler selects "other" as a reason, wranglers can be presented the option to create a new reason.)

5. How should we avoid race conditions? Should we simply prevent a wrangler from marking a crop rejected if it's already approved and vice versa - with a descriptive error message?

6. Should it be impossible to navigate to rejected crop by going to it's url directly? Should rejected crops be listed anywhere?

### CODE REVIEW

File: app/controllers/crops_controller.rb [hunk 1]
```diff
@@ -32,7 +32,8 @@ def index
 
   # GET /crops/wrangle
   def wrangle
-    @crops = Crop.recent.paginate(:page => params[:page])
+    @recent_crops = Crop.recent.paginate(:page => params[:page])
+    @pending_approval = Crop.pending_approval
```
Comment (tygriffin) [inline app/controllers/crops_controller.rb]: Note to self: This should be paginated as well.

File: app/controllers/crops_controller.rb [hunk 2]
```diff
@@ -120,8 +135,16 @@ def create
...
 
+    previous_status = @crop.approval_status
+
     respond_to do |format|
       if @crop.update(crop_params)
+        if previous_status == "pending"
+          requester = @crop.requester
+          new_status = @crop.approval_status
+          Notifier.crop_request_approved(requester, @crop).deliver! if new_status == "approved"
+          Notifier.crop_request_rejected(requester, @crop).deliver! if new_status == "rejected"
```
Comment (tygriffin) [inline app/controllers/crops_controller.rb]: Note to self: If I refactor this using `send` to get rid duplication will that be annoying and less explicit?

File: spec/mailers/notifier_spec.rb [hunk 3]
```diff
@@ -44,4 +44,84 @@
...
+    end
+
+    it 'includes the approved crop URL' do
+      expect(mail.body.encoded).to match crop_url(crop)
+    end
+
+    it 'includes links to plant, harvest and stash seeds for the new crop' do
+      expect(mail.body.encoded).to match new_planting_url
+      expect(mail.body.encoded).to match new_harvest_url
+      expect(mail.body.encoded).to match new_seed_url
```
Comment (tygriffin) [inline spec/mailers/notifier_spec.rb]: How can I check that the query parameter is included as well? The following wasn't working for some reason:

`expect(mail.body.encoded).to match new_planting_url(crop_id: crop.id)`

Can we just use `have_link` in this test?

File: app/controllers/crops_controller.rb (hunk 2 above)
Comment (tygriffin) [inline app/controllers/crops_controller.rb]: Note to self: If approved set the currently logged in wrangler as the creator.

File: app/models/crop.rb [hunk 4]
```diff
@@ -21,16 +22,18 @@ class Crop < ActiveRecord::Base
   before_destroy {|crop| crop.posts.clear}
 
   default_scope { order("lower(name) asc") }
```
Comment (Skud) [inline app/models/crop.rb]: Can we put the where(:approval_status => "approved") into the default scope and save ourselves having to repeat it many times below?  Then when we want the pending approval ones, we just ask for those.

File: app/models/crop.rb [hunk 5]
```diff
@@ -21,16 +22,18 @@ class Crop < ActiveRecord::Base
...
+  scope :randomized, -> { where(:approval_status => "approved").reorder('random()') } # ok on sqlite and psql, but not on mysql
+  scope :pending_approval, -> { where(:approval_status => "pending") }
 
   validates :en_wikipedia_url,
     :format => {
       :with => /\Ahttps?:\/\/en\.wikipedia\.org\/wiki/,
       :message => 'is not a valid English Wikipedia URL'
-    }
+    },
+    :if => :approved?
```
Comment (Skud) [inline app/models/crop.rb]: maybe leave a comment here explaining the :if => approved thing

**Comment by Skud** [comment]
Feedback:

1. Is the language natural enough for regular members on the new crops form page?

Can we see a screenshot?  I think it's probably ok though.

2. Should we use Javascript to pop up the reason for rejection only when approval_status is "rejected" on the edit crops form?

If you feel like it. I wouldn't say it's required though. We can expect crop wranglers to handle it okay, I think.

3. Is the reason_for_rejection required if rejected?

Yes.

4. Should the reasons for rejection be hardcoded in the model or should they be in a database table? (Database seems to make more sense. That way, if a wrangler selects "other" as a reason, wranglers can be presented the option to create a new reason.)

Would we ever want to look up all the crops that were denied for X reason? If so, database table. I can't imagine wanting it though.

Here are some suggestions for reasons, to start with:

* already in database
* not edible
* not enough information
* other

5.  How should we avoid race conditions? Should we simply prevent a wrangler from marking a crop rejected if it's already approved and vice versa - with a descriptive error message?

Yes, I think so. Also if they visit the edit page and it's already been approved they should clearly see that that has happened.  I think this implies we might want to include some metadata for wranglers on the edit page, eg. "requested by X on date. approved/rejected by Y on date."

6. Should it be impossible to navigate to rejected crop by going to it's url directly? Should rejected crops be listed anywhere?

Yes, and yes -- they should be listed on the crop wrangler page somewhere (or on a subpage). Nowhere else for now, though, I think.

File: app/models/crop.rb (hunk 4 above)
Comment (tygriffin) [inline app/models/crop.rb]: I wasn't sure exactly how to do this. If I do this 
```
default_scope { where(:approval_status => "approved").reorder("created_at desc")  }
scope :pending_approval, -> { where(:approval_status => "pending") }
```

I get queries like this:

`SELECT "crops".* FROM "crops"  WHERE "crops"."approval_status" = 'approved' AND "crops"."approval_status" = 'pending'  ORDER BY lower(name) asc`

**Comment by tygriffin** [comment]
**Ready for review**

A couple things may need some additional discussion:

1. Is this language good for non-wrangler members?
screen shot 2015-02-04 at 2 44 48 pm

2. Do we want to include any more standard reasons for rejection other than "already in database", "not edible", "not enough information", and "other"?

**Comment by Skud** [comment]
1. Language to use: I would suggest showing different language to crop wranglers vs non-wranglers.  The wranglers can see what we already have.  For non-wranglers, I'd put something like 

* Please provide the common name for the crop, in English, if you know it.
* Please provide a link to the crop's page on the English language Wikipedia. Crops without Wikipedia pages cannot be added to our database at this time.

2. I think that'll do for now and we can see what people put in the comments for "other" if there are a lot of them.  If you go through the discussion thread at http://growstuff.org/posts/skud-20130319-requests-for-new-crops you can see the reasons given in the past. The ones I see there are:

* silverbeet - already in db
* kumara - already in db
* rhubarb - already in db
* spider plant - not edible
* rabbit foot fern - not edible
* madder - not edible
* longish list of cover crops etc - not edible
* daisy - other (insufficiently specific; some are edible and some aren't)
* rocket - already in db
* cerejeira - already in db
* peace lily - not edible

And lots and lots that were "there's no wikipedia page for this so we can't add it" which is basically "not enough information" I guess?

**Comment by Skud** [comment]
Just thinking about this a bit more and I think the page title should vary too:

* Non-wrangler: "Suggest a crop"
* Wrangler, reviewing a suggested crop: "Review crop: $cropname"
* Wrangler, adding a brand new crop: "New crop"

Also there should be some information for non-wranglers, to help them understand the process.  At the top of the page:

> Thanks for taking the time to suggest a crop! Our crop database is managed by volunteers, and we appreciate your help.  Here are some things to consider when suggesting a new crop:
> 
> * First, you might want to [search our crops](http://growstuff.org/crops/search) to make sure we don't have it already, perhaps under an alternate name.
> * The Growstuff database only contains edible crops.  In future we hope to support other crops, but for now, if your suggestion is not edible we won't be able to add it.
> * At this time, we are only adding crops which have a Wikipedia page. If you want to add a specific variety of crop that doesn't have its own Wikipedia entry, please use the more general form of the crop instead and put the name of your variety in the notes/description.

(Note: that crop search link doesn't actually include a search form, except in the navbar. We might want to include one there just so we can point people at it.)

After the form but before the submit button:

> When you submit this form, your suggestion will be sent to our team of [volunteer crop wranglers](http://talk.growstuff.org/c/crop-wrangling) for review.  We'll let you know the outcome as soon as we can.

**Comment by tygriffin** [comment]
Good ideas! 

I didn't add the second crop search field to `/crops/search` because I thought the one in the nav bar was pretty obvious. What do you think though? Am I wrong on that one?

**Comment by tygriffin** [comment]
Also, we're starting to accumulate a lot of logic in the views. What do you think about making a new story to add decorators to Growstuff a la [Draper](https://github.com/drapergem/draper)

(PS: That gem is looking for a new maintainer so maybe not so great, but we could make our own light-weight decorator class.)

**Comment by Skud** [comment]
Draper looks awesome but let's hold off on that for now.

**Comment by Skud** [comment]
@tygriffin re: search, the problem is that http://staging.growstuff.org/crops/search (linking to staging since that has elasticsearch) is a really unfriendly, weird page to link someone to.  The heading is weird and ungrammatical (I mentioned this in the "please test" thread for elasticsearch) and then it says "no results found, please do this other thing".  I am not thrilled about linking to that page.  It's not the end of the world but I would like to fix it up.  I guess we could fork a separate story.

**Pull request merged**